File: python/trees/validate-binary-search-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def isValidBST(self, root: Optional[TreeNode]) -> bool:
        def dfs(node):
            if not node:
                return None, None, True

            leftmin, leftmax, leftvalid = dfs(node.left)
            rightmin, rightmax, rightvalid = dfs(node.right)

            if (leftmin == None or node.val > leftmax) and leftvalid and (
                    rightmax == None or node.val < rightmin) and rightvalid:
                if not leftmin and not rightmax:
                    return node.val, node.val, True
                if not leftmin:
                    return min(node.val, rightmin), max(node.val, rightmax), True
                if not rightmax:
                    return min(node.val, leftmin), max(node.val, leftmax), True

                return min(leftmin, rightmin, node.val), max(leftmax, rightmax, node.val), True

            return 0, 0, False

        _, _, answer = dfs(root)
        return answer

File: c++/trees/validate-binary-search-tree.cc
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    bool isValidBST(TreeNode* root) {
        return validate(root, std::numeric_limits<long long>::min(), std::numeric_limits<long long>::max());
    }

private:
    bool validate(TreeNode* node, long long minimum, long long maximum) {
        if (!node) {
            return true;
        }

        return node->val > minimum && node->val < maximum && validate(node->left, minimum, node->val) && validate(node->right, node->val, maximum);
    }
};
